### PR TITLE
use default ip for advertise URL

### LINKED ISSUE #2858: etcd should use default values for advertised client and peer URLs.
I tried configuring etcd v2.010 with the following settings:

```
ETCD_NAME=etcd
ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379
ETCD_LISTEN_PEER_URLS=http://0.0.0.0:2380
ETCD_DATA_DIR=/var/lib/etcd
ETCD_INITIAL_CLUSTER_TOKEN=kubernetes
```

I got the following error:

```
May 22 13:07:04 localhost systemd[1]: Started etcd.
May 22 13:07:04 localhost systemd[1]: Starting etcd...
May 22 13:07:04 localhost systemd[1]: etcd.service: main process exited, code=exited, status=1/FAILURE
May 22 13:07:04 localhost systemd[1]: Unit etcd.service entered failed state.
May 22 13:07:04 localhost systemd[1]: etcd.service failed.
May 22 13:07:04 localhost etcd[6165]: 2015/05/22 13:07:04 error verifying flags, -advertise-client-urls is required when -listen-client-urls is set explicitly. See 'etcd -help'.
May 22 13:07:04 localhost etcd[6165]: 2015/05/22 13:07:04 When listening on specific address(es), this etcd process must advertise accessible url(s) to each connected client.
```

How I expect this to work:

I'm currently trying to bootstrap a single node etcd cluster, and don't know the IP address ahead of time. Ideally I want to listen to all interfaces and allow clients such as Kubernetes and etcdctl to interact with etcd.

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #2858

### CODE REVIEW

**Comment by gyuho** [comment]
Is there anyway that I can try `GetDefaultHost`? In my linux machine, I get

```
could not find default route
```

**Comment by heyitsanthony** [comment]
@gyuho huh, works on my machine. Is it failing to get `defaultNlMsg` or is it failing to find the `RTA_PREFSRC`? I'll see if it works on my ubuntu machine...

**Comment by gyuho** [comment]
@heyitsanthony It's returning here

```go
...
	for _, attr := range attrs {
		if attr.Attr.Type == syscall.RTA_PREFSRC {
			return net.IP(attr.Value).String(), nil
		}
	}

	// returns here
	return "", errNoDefaultRoute
}
```

```
go version go1.7rc6 linux/amd64

Linux gyuho 4.4.0-34-generic #53-Ubuntu SMP Wed Jul 27 16:06:39 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by heyitsanthony** [comment]
@gyuho OK, could reproduce on my ubuntu machine; can work around. Thanks for giving this a spin!

**Comment by heyitsanthony** [comment]
had to query the iface to get the address on ubuntu /cc @gyuho

**Comment by gyuho** [comment]
Ok works now. Mine is

```
10.0.0.30
```

LGTM. /cc @xiang90

File: embed/config.go [hunk 1]
```diff
@@ -50,8 +47,25 @@ var (
...
+	DefaultAdvertiseClientURLs      = "http://localhost:2379"
 )
 
+func init() {
+	ip, err := netutil.GetDefaultHost()
+	if err != nil {
+		return
+	}
+	// found default host; listen everywhere, advertise on default host
+	DefaultListenPeerURLs = "http://0.0.0.0:2380"
```
Comment (xiang90) [inline embed/config.go]: i do not think we should listen on wide by default even if we found the default host for security reasons. by finding default adv url, users do not need to set avd ulrs themselves but not listen url I think.

File: pkg/netutil/routes.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// +build !linux !386,!amd64
+
+package netutil
+
+import "fmt"
+
+// GetDefaultHost fetches the a resolvable name that corresponds
+// to the machine's default routable interface
+func GetDefaultHost() (string, error) {
+	return "", fmt.Errorf("default host not supported on this platform")
```
Comment (xiang90) [inline pkg/netutil/routes.go]: on %v, os.GOARCH

File: embed/config.go [hunk 3]
```diff
@@ -50,8 +47,25 @@ var (
...
+	DefaultListenPeerURLs           = "http://localhost:2380"
+	DefaultListenClientURLs         = "http://localhost:2379"
+	DefaultInitialAdvertisePeerURLs = "http://localhost:2380"
+	DefaultAdvertiseClientURLs      = "http://localhost:2379"
 )
 
+func init() {
+	ip, err := netutil.GetDefaultHost()
+	if err != nil {
+		return
```
Comment (xiang90) [inline embed/config.go]: print out error if err is equal to errNoDefaultRoute?

File: pkg/netutil/routes_linux.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build 386 amd64
+
+// TODO support native endian but without using "unsafe"
```
Comment (xiang90) [inline pkg/netutil/routes_linux.go]: where do we use unsafe currently?

File: pkg/netutil/routes_linux.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	rtmsg := syscall.RtMsg{}
+	for _, m := range msgs {
+		if m.Header.Type != syscall.RTM_NEWROUTE {
+			continue
+		}
+		buf := bytes.NewBuffer(m.Data[:syscall.SizeofRtMsg])
+		if rerr := binary.Read(buf, binary.LittleEndian, &rtmsg); rerr != nil {
+			continue
+		}
+		if rtmsg.Dst_len == 0 {
```
Comment (xiang90) [inline pkg/netutil/routes_linux.go]: Add comments?

//  For RTM_GETROUTE,  setting rtm_dst_len and rtm_src_len to 0 means you get all
// entries for the specified routing table. 

probably also check src_len?

Comment (xiang90) [inline pkg/netutil/routes_linux.go]: Hmm... ignore previous comment. that is for request...

File: pkg/netutil/routes_linux.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	msgs, msgErr := syscall.ParseNetlinkMessage(dat)
+	if msgErr != nil {
+		return nil, msgErr
+	}
+
+	rtmsg := syscall.RtMsg{}
+	for _, m := range msgs {
+		if m.Header.Type != syscall.RTM_NEWROUTE {
```
Comment (xiang90) [inline pkg/netutil/routes_linux.go]: is this doced somewhere? the response type is the same to the request type (create a new route)?

File: embed/config.go [hunk 7]
```diff
@@ -50,8 +47,25 @@ var (
...
+	}
+	// found default host; listen everywhere, advertise on default host
```
Comment (xiang90) [inline embed/config.go]: print out that we use ip as default after finding default host?

File: pkg/netutil/routes_linux.go (hunk 4 above)
Comment (heyitsanthony) [inline pkg/netutil/routes_linux.go]: we don't afaik, but the only way to get the native endian I'm aware of uses `unsafe`, hence the comment

Comment (xiang90) [inline pkg/netutil/routes_linux.go]: oh... ok.

File: pkg/netutil/routes_linux.go (hunk 5 above)
Comment (heyitsanthony) [inline pkg/netutil/routes_linux.go]: ok, added comment about dst_len == 0 meaning default route

File: pkg/netutil/routes_linux.go (hunk 6 above)
Comment (heyitsanthony) [inline pkg/netutil/routes_linux.go]: It's just the netlink protocol, I don't think it needs documentation?

File: pkg/netutil/routes.go (hunk 2 above)
Comment (heyitsanthony) [inline pkg/netutil/routes.go]: using GOOS_GOARCH

File: embed/config.go (hunk 1 above)
Comment (heyitsanthony) [inline embed/config.go]: ok

File: embed/config.go (hunk 3 above)
Comment (heyitsanthony) [inline embed/config.go]: added error reporting to `etcdmain`

File: embed/config.go (hunk 7 above)
Comment (heyitsanthony) [inline embed/config.go]: added to `etcdmain`; will give a notification if the config uses the default

**Comment by heyitsanthony** [comment]
@xiang90 all fixed up

File: etcdmain/etcd.go [hunk 8]
```diff
@@ -184,6 +184,15 @@ func startEtcdOrProxyV2() {
 
 // startEtcd runs StartEtcd in addition to hooks needed for standalone etcd.
 func startEtcd(cfg *embed.Config) (<-chan struct{}, <-chan error, error) {
+	defaultHost, dhErr := cfg.IsDefaultHost()
+	if defaultHost != "" {
+		if dhErr == nil {
+			plog.Infof("advertising using default host %q", defaultHost)
```
Comment (xiang90) [inline etcdmain/etcd.go]: using detected default host

**Comment by xiang90** [comment]
lgtm

**Pull request merged**